pkg/cluster/self_hosted.go

```diff
@@ -113,3 +114,93 @@ func (c *Cluster) newSelfHostedSeedMember() error {
 	c.logger.Infof("self-hosted cluster created with seed member (%s)", newMemberName)
 	return nil
 }
+
+func (c *Cluster) migrateBootMember() error {
+	endpoint := c.spec.SelfHosted.BootMemberClientEndpoint
+
+	c.logger.Infof("migrating boot member (%s)", endpoint)
+
+	cfg := clientv3.Config{
+		Endpoints:   []string{endpoint},
+		DialTimeout: constants.DefaultDialTimeout,
+	}
+	etcdcli, err := clientv3.New(cfg)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.MemberList(ctx)
+	cancel()
+	etcdcli.Close()
+	if err != nil {
+		return err
+	}
+
+	if len(resp.Members) != 1 {
+		return fmt.Errorf("boot cluster contains more than one member")
+	}
+	bootMember := resp.Members[0]
+	bootMemberID := resp.Members[0].ID
+
+	initialCluster := make([]string, 0)
+	for _, purl := range bootMember.PeerURLs {
+		initialCluster = append(initialCluster, fmt.Sprintf("%s=%s", bootMember.Name, purl))
+	}
+
+	// create the first member inside Kubernetes for migration
+	newMemberName := fmt.Sprintf("%s-%04d", c.name, c.idCounter)
+	c.idCounter++
+
+	peerURL := "http://$(MY_POD_IP):2380"
+	initialCluster = append(initialCluster, newMemberName+"="+peerURL)
+
+	pod := k8sutil.MakeSelfHostedEtcdPod(newMemberName, initialCluster, c.name, "existing", "", c.spec)
+	pod = k8sutil.PodWithAddMemberInitContainer(pod, []string{endpoint}, newMemberName, []string{peerURL}, c.spec)
+
+	err = k8sutil.CreateAndWaitPod(c.kclient, c.namespace, pod, 30*time.Second)
+	if err != nil {
+		return err
+	}
+
+	// update pod status to get its IP
+	pod, err = c.kclient.Pods(c.namespace).Get(pod.Name)
+	if err != nil {
+		return err
+	}
+
+	delay := 20 * time.Second
+	c.logger.Infof("wait %v before removing the boot member", delay)
+	time.Sleep(delay)
+
+	cfg = clientv3.Config{
+		Endpoints:   []string{"http://" + pod.Status.PodIP + ":2379"},
+		DialTimeout: constants.DefaultDialTimeout,
+	}
+	etcdcli, err = clientv3.New(cfg)
+	if err != nil {
+		return err
+	}
+
+	// delete the boot member from the etcd cluster.
+	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	_, err = etcdcli.Cluster.MemberRemove(ctx, bootMemberID)
+	cancel()
+	if err != nil {
+		return fmt.Errorf("etcdcli failed to remove boot member: %v", err)
+	}
+
+	// TODO: do not wait forever.
+	for {
```

Comment:
We should probably stop depending on external util funcs. it is small, but when it breaks (like update dependencies) it causes problem to us. if we do want this kind of thing, we should write our own.